lib/msal-node/src/client/ManagedIdentityApplication.ts

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+    AccessTokenEntity,
+    AuthenticationResult,
+    Authority,
+    AuthorityOptions,
+    CacheManager,
+    ClientAuthErrorCodes,
+    Constants,
+    CredentialFilter,
+    CredentialType,
+    DEFAULT_CRYPTO_IMPLEMENTATION,
+    DefaultStorageClass,
+    INetworkModule,
+    Logger,
+    NetworkManager,
+    ProtocolMode,
+    ResponseHandler,
+    ScopeSet,
+    TimeUtils,
+    createClientAuthError,
+} from "@azure/msal-common";
+import { ServiceFabric } from "./ManagedIdentitySources/ServiceFabric";
+import { AppService } from "./ManagedIdentitySources/AppService";
+import { CloudShell } from "./ManagedIdentitySources/CloudShell";
+import { AzureArc } from "./ManagedIdentitySources/AzureArc";
+import { Imds } from "./ManagedIdentitySources/Imds";
+import {
+    ManagedIdentityConfiguration,
+    ManagedIdentityNodeConfiguration,
+    buildManagedIdentityConfiguration,
+} from "../config/Configuration";
+import { version, name } from "../packageMetadata.js";
+import { ManagedIdentityRequest } from "../request/ManagedIdentityRequest";
+import { ManagedIdentityId } from "../config/ManagedIdentityId";
+import { CryptoProvider } from "../crypto/CryptoProvider";
+
+/**
+ * Class to initialize a managed identity and identify the service
+ */
+export class ManagedIdentityApplication {
+    private managedIdentityId: ManagedIdentityId;
+
+    private config: ManagedIdentityNodeConfiguration;
+
+    private logger: Logger;
+    private cacheManager: CacheManager;
+    private networkClient: INetworkModule;
+    private networkManager: NetworkManager;
+    private cryptoProvider: CryptoProvider;
+
+    private identitySource:
+        | ServiceFabric
+        | AppService
+        | CloudShell
+        | AzureArc
+        | Imds;
+
+    private fakeAuthority: Authority;
+
+    constructor(
+        managedIdentityId: ManagedIdentityId,
+        configuration?: ManagedIdentityConfiguration
+    ) {
+        this.config = buildManagedIdentityConfiguration(configuration || {});
+        this.managedIdentityId = managedIdentityId;
+
+        this.logger = new Logger(
+            this.config.system.loggerOptions,
+            name,
+            version
+        );
+
+        this.cacheManager = new DefaultStorageClass(
+            this.managedIdentityId.id,
+            DEFAULT_CRYPTO_IMPLEMENTATION,
+            this.logger
+        );
+
+        this.networkClient = this.config.system.networkClient;
+        this.networkManager = new NetworkManager(
+            this.networkClient,
+            this.cacheManager
+        );
+
+        this.cryptoProvider = new CryptoProvider();
+
+        this.identitySource = this.selectManagedIdentitySource();
```

Comment:
We shouldn't try to select the MI source when creating the app, but when firing the request.
Once a request has figured out what MI source we're on (e.g. VM), all other requests need to remeber this (i.e. static cache this)

Reply:
I disagree. If we select the MSI source when creating the app, then it will already be known when firing the requests. Isn't this how .dotnet does it?